File: 2904ShortestAndLexicographicallySmallestBeatifulString.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:
        ans = ""
        ones_count = 0
        left = 0

        for right in range(len(s)):
            if s[right] == "1":
                ones_count += 1
            print("right", right, ones_count)
            while ones_count == k:
                sub = s[left : right + 1]
                print("right", right, ones_count, sub)
                if (
                    not ans
                    or len(sub) < len(ans)
                    or (len(sub) == len(ans) and sub < ans)
                ):
                    ans = sub

                if s[left] == "1":
                    ones_count -= 1
                left += 1

        return ans


solution = Solution()
print(solution.shortestBeautifulSubstring("100011001", 3))
# print(solution.shortestBeautifulSubstring("000", 3))

File: 2904ShortestAndLexicographicallySmallestBeatifulString.java

class Solution {

    public String shortestBeautifulSubstring(String s, int k) {
        String ans = "";
        int onesCount = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            if (s.charAt(right) == '1') {
                onesCount++;
            }

            while (onesCount == k) {
                String sub = s.substring(left, right + 1);

                if (ans.isEmpty() || sub.length() < ans.length()
                        || (sub.length() == ans.length() && sub.compareTo(ans) < 0)) {
                    ans = sub;
                }

                if (s.charAt(left) == '1') {
                    onesCount--;
                }
                left++;
            }
        }
        return ans;
    }
}
